### ISSUE TITLE
Issue with 'Key' Being a Required Property in cfn-lint v1.9.3

### ISSUE THREAD

**Issue by wapacz** [aws-cloudformation/cfn-lint#3562]
### CloudFormation Lint Version

1.9.3

### What operating system are you using?

Ubuntu

### Describe the bug

I have following part of the template that latest (1.9.3) cfn-lint complains about:

```
      Tags:

        - !If
          - Condition1
          - Key: 'Key1'
            Value: 'Value1'
          - !Ref 'AWS::NoValue'
```

I'm getting following error regarding this part: `3024 'Key' is a required property`

Previous version 1.9.2 version didn't complain about it.

### Expected behavior

The template should work as it is, or please clarify how this should be properly implemented.

### Reproduction template

```yaml
---
AWSTemplateFormatVersion: 2010-09-09
Parameters:
  Property1:
    Type: 'String'
    Default: 'no'

Conditions:
  Condition1: !Equals [!Ref 'Property1', 'yes']

Resources:
  Resource1:
    Type: 'AWS::IAM::Role'
    Properties:
      RoleName: 'Role1'
      AssumeRolePolicyDocument:
        Version: 2012-10-17
      Tags:

        - !If
          - Condition1
          - Key: 'Key1'
            Value: 'Value1'
          - !Ref 'AWS::NoValue'

        - Key: 'Key2'
          Value: 'Value2'
```

**Comment by nosnilmot** [comment]
I think the problem is `AWS::NoValue` is not recognised as a (valid) dummy value in a list where an object is expected.

An even simpler reproducer:
```yaml
AWSTemplateFormatVersion: '2010-09-09'
Resources:
  SecurityGroup:
    Type: AWS::EC2::SecurityGroup
    Properties:
      GroupDescription: Security Group
      VpcId: 'vpc-12345678901234567'
      Tags:
        - !Ref 'AWS::NoValue'
```

**Comment by wapacz** [comment]
@nosnilmot, thanks. Yes, that is true. It looks like problem is with `AWS::NoValue` on the list.

**Comment by kddejong** [comment]
Correct it has to do with how we process `items`.  It creates an interesting scenario that I'm working to resolve.

For instance in the scenario below we want to have the error `E3003 'AssumeRolePolicyDocument' is a required property`

```
---
Resources:
  Resource2:
    Type: 'AWS::IAM::Role'
    Properties: !Ref AWS::NoValue
```

But for in a list of objects we want the list item removed as the following template should provide 0 errors (at least for the tags).

```
---
AWSTemplateFormatVersion: 2010-09-09
Parameters:
  Property1:
    Type: 'String'
    Default: 'no'

Conditions:
  Condition1: !Equals [!Ref 'Property1', 'yes']

Resources:
  Resource1:
    Type: 'AWS::IAM::Role'
    Properties:
      Tags:

        - !If
          - Condition1
          - Key: 'Key1'
            Value: 'Value1'
          - !Ref 'AWS::NoValue'

        - Key: 'Key2'
          Value: 'Value2'
        - !Ref 'AWS::NoValue'
```

In both scenarios there is a required keyword but in the list side of the world we need to recognize it is not an item in the list at all.  As a result of this I'm going to send the `items` json schema word through the same logic we use to determine minItems, maxItems, maxProperties, etc. as defined [here](https://github.com/aws-cloudformation/cfn-lint/blob/b16d6b4ce0796020a0ad5ee1c56453fde6079f2b/src/cfnlint/jsonschema/_filter.py#L35-L53)

**Comment by kddejong** [comment]
Additionally this change will handle all these scenarios appropriately

```yaml
---
AWSTemplateFormatVersion: 2010-09-09
Parameters:
  Property1:
    Type: 'String'
    Default: 'no'

Conditions:
  Condition1: !Equals [!Ref 'Property1', 'yes']

Resources:
  Resource1:
    Type: 'AWS::IAM::Role'
    Properties:
      AssumeRolePolicyDocument: {}
      Tags:

        - !If
          - Condition1
          - Key: 'Key1'
            Value: 'Value1'
          - !Ref 'AWS::NoValue'

        - Key: 'Key2'
          Value: 'Value2'
        - !Ref 'AWS::NoValue'
  Resource2:
    Type: 'AWS::IAM::Role'
    Properties: !Ref AWS::NoValue
  Resource3:
    Type: 'AWS::IAM::Role'
    Properties:
      AssumeRolePolicyDocument: {}
      Tags:
        !If
          - Condition1
          - - Key: 'Key1'
              Value: 'Value1'
          - !Ref 'AWS::NoValue'
```

```
E3003 'AssumeRolePolicyDocument' is a required property
local/issue/3562.yaml:29:5
```
